sequenceT in v1.15.0 causes type exceptions

Caused by #766 

It seems the new mapped types for sequenceT cause exceptions when working directly with the unknown type `F` - verified does not cause this exception in `v1.14.4`

Some suggestions to resolve:

1) Restore the previous interface for `sequenceT` but add a final case that falls back upon Mapped Tuple Types
2) Add `sequenceT2`, `sequenceT3`, or `sequenceTn` methods to utilize previous style

## Example
_Error caused in `factory` during call to `create`_

```ts
import { Apply1, sequenceT } from 'fp-ts/lib/Apply'
import { log } from 'fp-ts/lib/Console'
import { Type, URIS } from 'fp-ts/lib/HKT'
import { IO, io } from 'fp-ts/lib/IO'

export interface Entity<TData>
    extends Readonly<{
        uri: [string, string]
        data: TData
        createdAt: Date
        updatedAt: Date
    }> {}

export type Uri = [string, string]

export interface UserData extends Readonly<{ email: string }> {}
export interface User extends Entity<UserData> {}

export type Factory<F extends URIS, T, R> = (data: T) => Type<F, R>

export const uri = (domain: string, id: string): Uri => [domain, id]

export const create = <TData>(uri: Uri, data: TData, t: Date): Entity<TData> => ({
    uri,
    data,
    createdAt: t,
    updatedAt: t
})

export const factory = <F extends URIS>(F: Apply1<F>, fDate: Type<F, Date>) => <TData>(
    fUri: Type<F, Uri>
): Factory<F, TData, Entity<TData>> => (data: TData) =>
    F.map(sequenceT(F)(fUri, fDate), ([uri, t]) => create(uri, data, t))

const ioTestDate = new IO(() => new Date('2018-01-01T00:00:00Z'))
const ioTestEntity = factory(io, ioTestDate)

const ioTestUserUri = new IO(() => uri('user', 'user1'))
const testUserFactory: Factory<'IO', { email: string }, User> = ioTestEntity(ioTestUserUri)

testUserFactory({ email: 'me@example.com' })
    .chain(log)
    .run()
```

## Expected Output 
```
{ uri: [ 'user', 'user1' ],
  data: { email: 'me@example.com' },
  createdAt: 2018-01-01T00:00:00.000Z,
  updatedAt: 2018-01-01T00:00:00.000Z }
```

## Actual Result
```
src/example.ts(33,59): error TS2345: Argument of type 'Type<F, [string, string]> extends Type<F, infer A> ? A : never' is not assignable to parameter of type '[string, string]'.
  Type '{}' is missing the following properties from type '[string, string]': 0, 1, length, pop, and 28 more.
```

--- a/docs/modules/Apply.ts.md
+++ b/docs/modules/Apply.ts.md
@@ -27,6 +27,12 @@ Formally, `Apply` represents a strong lax semi-monoidal endofunctor.
 - [Apply2C (interface)](#apply2c-interface)
 - [Apply3 (interface)](#apply3-interface)
 - [Apply3C (interface)](#apply3c-interface)
+- [SequenceT (interface)](#sequencet-interface)
+- [SequenceT1 (interface)](#sequencet1-interface)
+- [SequenceT2 (interface)](#sequencet2-interface)
+- [SequenceT2C (interface)](#sequencet2c-interface)
+- [SequenceT3 (interface)](#sequencet3-interface)
+- [SequenceT3C (interface)](#sequencet3c-interface)
 - [applyFirst (function)](#applyfirst-function)
 - [applySecond (function)](#applysecond-function)
 - [getSemigroup (function)](#getsemigroup-function)
@@ -97,6 +103,90 @@ export interface Apply3<F extends URIS3> extends Functor3<F> {
 ```ts
 export interface Apply3C<F extends URIS3, U, L> extends Functor3C<F, U, L> {
   readonly ap: <A, B>(fab: Type3<F, U, L, (a: A) => B>, fa: Type3<F, U, L, A>) => Type3<F, U, L, B>
+}
+```
+
+# SequenceT (interface)
+
+**Signature**
+
+```ts
+export interface SequenceT<F> {
+  <T extends Array<HKT<F, any>>>(...t: T & { 0: HKT<F, any> }): HKT<
+    F,
+    { [K in keyof T]: [T[K]] extends [HKT<F, infer A>] ? A : never }
+  >
+}
+```
+
+# SequenceT1 (interface)
+
+**Signature**
+
+```ts
+export interface SequenceT1<F extends URIS> {
+  <T extends Array<Type<F, any>>>(...t: T & { 0: Type<F, any> }): Type<
+    F,
+    { [K in keyof T]: [T[K]] extends [Type<F, infer A>] ? A : never }
+  >
+}
+```
+
+# SequenceT2 (interface)
+
+**Signature**
+
+```ts
+export interface SequenceT2<F extends URIS2> {
+  <L, T extends Array<Type2<F, L, any>>>(...t: T & { 0: Type2<F, L, any> }): Type2<
+    F,
+    L,
+    { [K in keyof T]: [T[K]] extends [Type2<F, L, infer A>] ? A : never }
+  >
+}
+```
+
+# SequenceT2C (interface)
+
+**Signature**
+
+```ts
+export interface SequenceT2C<F extends URIS2, L> {
+  <T extends Array<Type2<F, L, any>>>(...t: T & { 0: Type2<F, L, any> }): Type2<
+    F,
+    L,
+    { [K in keyof T]: [T[K]] extends [Type2<F, L, infer A>] ? A : never }
+  >
+}
+```
+
+# SequenceT3 (interface)
+
+**Signature**
+
+```ts
+export interface SequenceT3<F extends URIS3> {
+  <U, L, T extends Array<Type3<F, U, L, any>>>(...t: T & { 0: Type3<F, U, L, any> }): Type3<
+    F,
+    U,
+    L,
+    { [K in keyof T]: [T[K]] extends [Type3<F, U, L, infer A>] ? A : never }
+  >
+}
+```
+
+# SequenceT3C (interface)
+
+**Signature**
+
+```ts
+export interface SequenceT3C<F extends URIS3, U, L> {
+  <T extends Array<Type3<F, U, L, any>>>(...t: T & { 0: Type3<F, U, L, any> }): Type3<
+    F,
+    U,
+    L,
+    { [K in keyof T]: [T[K]] extends [Type3<F, U, L, infer A>] ? A : never }
+  >
 }
 ```
 
@@ -308,28 +398,32 @@ export function sequenceS<F extends URIS3>(
   F: Apply3<F>
 ): <U, L, R extends Record<string, Type3<F, U, L, any>>>(
   r: EnforceNonEmptyRecord<R> & Record<string, Type3<F, U, L, any>>
-) => Type3<F, U, L, { [K in keyof R]: R[K] extends Type3<F, any, any, infer A> ? A : never }>
+) => Type3<F, U, L, { [K in keyof R]: [R[K]] extends [Type3<F, any, any, infer A>] ? A : never }>
 export function sequenceS<F extends URIS3, U, L>(
   F: Apply3C<F, U, L>
 ): <R extends Record<string, Type3<F, U, L, any>>>(
   r: EnforceNonEmptyRecord<R>
-) => Type3<F, U, L, { [K in keyof R]: R[K] extends Type3<F, any, any, infer A> ? A : never }>
+) => Type3<F, U, L, { [K in keyof R]: [R[K]] extends [Type3<F, any, any, infer A>] ? A : never }>
 export function sequenceS<F extends URIS2>(
   F: Apply2<F>
 ): <L, R extends Record<string, Type2<F, L, any>>>(
   r: EnforceNonEmptyRecord<R> & Record<string, Type2<F, L, any>>
-) => Type2<F, L, { [K in keyof R]: R[K] extends Type2<F, any, infer A> ? A : never }>
+) => Type2<F, L, { [K in keyof R]: [R[K]] extends [Type2<F, any, infer A>] ? A : never }>
 export function sequenceS<F extends URIS2, L>(
   F: Apply2C<F, L>
 ): <R extends Record<string, Type2<F, L, any>>>(
   r: EnforceNonEmptyRecord<R>
-) => Type2<F, L, { [K in keyof R]: R[K] extends Type2<F, any, infer A> ? A : never }>
+) => Type2<F, L, { [K in keyof R]: [R[K]] extends [Type2<F, any, infer A>] ? A : never }>
 export function sequenceS<F extends URIS>(
   F: Apply1<F>
 ): <R extends Record<string, Type<F, any>>>(
   r: EnforceNonEmptyRecord<R>
-) => Type<F, { [K in keyof R]: R[K] extends Type<F, infer A> ? A : never }>
-export function sequenceS<F>(F: Apply<F>): (r: Record<string, HKT<F, any>>) => HKT<F, Record<string, any>> { ... }
+) => Type<F, { [K in keyof R]: [R[K]] extends [Type<F, infer A>] ? A : never }>
+export function sequenceS<F>(
+  F: Apply<F>
+): <R extends Record<string, HKT<F, any>>>(
+  r: EnforceNonEmptyRecord<R>
+) => HKT<F, { [K in keyof R]: [R[K]] extends [HKT<F, infer A>] ? A : never }> { ... }
 ```
 
 **Example**
@@ -365,36 +459,12 @@ Tuple sequencing, i.e., take a tuple of monadic actions and do them from left-to
 **Signature**
 
 ```ts
-export function sequenceT<F extends URIS3>(
-  F: Apply3<F>
-): <U, L, T extends Array<Type3<F, U, L, any>>>(
-  ...t: T & { 0: Type3<F, U, L, any> }
-) => Type3<F, U, L, { [K in keyof T]: T[K] extends Type3<F, U, L, infer A> ? A : never }>
-export function sequenceT<F extends URIS3, U, L>(
-  F: Apply3C<F, U, L>
-): <T extends Array<Type3<F, U, L, any>>>(
-  ...t: T & { 0: Type3<F, U, L, any> }
-) => Type3<F, U, L, { [K in keyof T]: T[K] extends Type3<F, U, L, infer A> ? A : never }>
-export function sequenceT<F extends URIS2>(
-  F: Apply2<F>
-): <L, T extends Array<Type2<F, L, any>>>(
-  ...t: T & { 0: Type2<F, L, any> }
-) => Type2<F, L, { [K in keyof T]: T[K] extends Type2<F, L, infer A> ? A : never }>
-export function sequenceT<F extends URIS2, L>(
-  F: Apply2C<F, L>
-): <T extends Array<Type2<F, L, any>>>(
-  ...t: T & { 0: Type2<F, L, any> }
-) => Type2<F, L, { [K in keyof T]: T[K] extends Type2<F, L, infer A> ? A : never }>
-export function sequenceT<F extends URIS>(
-  F: Apply1<F>
-): <T extends Array<Type<F, any>>>(
-  ...t: T & { 0: Type<F, any> }
-) => Type<F, { [K in keyof T]: T[K] extends Type<F, infer A> ? A : never }>
-export function sequenceT<F>(
-  F: Apply<F>
-): <T extends Array<HKT<F, any>>>(
-  ...t: T & { 0: HKT<F, any> }
-) => HKT<F, { [K in keyof T]: T[K] extends HKT<F, infer A> ? A : never }> { ... }
+export function sequenceT<F extends URIS3>(F: Apply3<F>): SequenceT3<F>
+export function sequenceT<F extends URIS3, U, L>(F: Apply3C<F, U, L>): SequenceT3C<F, U, L>
+export function sequenceT<F extends URIS2>(F: Apply2<F>): SequenceT2<F>
+export function sequenceT<F extends URIS2, L>(F: Apply2C<F, L>): SequenceT2C<F, L>
+export function sequenceT<F extends URIS>(F: Apply1<F>): SequenceT1<F>
+export function sequenceT<F>(F: Apply<F>): SequenceT<F> { ... }
 ```
 
 **Example**

--- a/dtslint/ts3.1/index.ts
+++ b/dtslint/ts3.1/index.ts
@@ -66,6 +66,35 @@ Apy.liftA2(applicativeValidation) // $ExpectType <A, B, C>(f: Curried2<A, B, C>)
 
 // sequenceS
 
+declare function functionForfactoryS(
+  a1: string,
+  a2: number,
+  a3: boolean,
+  a4: string,
+  a5: number,
+  a6: boolean,
+  a7: string,
+  a8: number,
+  a9: boolean
+): boolean
+
+export function factoryS<F extends H.URIS>(
+  F: Apy.Apply1<F>,
+  a1: H.Type<F, string>,
+  a2: H.Type<F, number>,
+  a3: H.Type<F, boolean>,
+  a4: H.Type<F, string>,
+  a5: H.Type<F, number>,
+  a6: H.Type<F, boolean>,
+  a7: H.Type<F, string>,
+  a8: H.Type<F, number>,
+  a9: H.Type<F, boolean>
+): H.Type<F, boolean> {
+  return F.map(Apy.sequenceS(F)({ a1, a2, a3, a4, a5, a6, a7, a8, a9 }), ({ a1, a2, a3, a4, a5, a6, a7, a8, a9 }) =>
+    functionForfactoryS(a1, a2, a3, a4, a5, a6, a7, a8, a9)
+  )
+}
+
 declare const sequenceS1: E.Either<string, number>
 declare const sequenceS2: E.Either<string, string>
 declare const sequenceS3: E.Either<string, boolean>
@@ -95,6 +124,23 @@ sequenceSf2({ sequenceS5, sequenceS9 })
 sequenceSf2({ sequenceS5, sequenceS6, sequenceS7 }) // $ExpectType ReaderTaskEither<{ a: number; }, string, { sequenceS5: number; sequenceS6: string; sequenceS7: boolean; }>
 
 // sequenceT
+
+export function factoryT<F extends H.URIS>(
+  F: Apy.Apply1<F>,
+  f1: H.Type<F, string>,
+  f2: H.Type<F, number>,
+  f3: H.Type<F, boolean>,
+  f4: H.Type<F, string>,
+  f5: H.Type<F, number>,
+  f6: H.Type<F, boolean>,
+  f7: H.Type<F, string>,
+  f8: H.Type<F, number>,
+  f9: H.Type<F, boolean>
+): H.Type<F, boolean> {
+  return F.map(Apy.sequenceT(F)(f1, f2, f3, f4, f5, f6, f7, f8, f9), ([a1, a2, a3, a4, a5, a6, a7, a8, a9]) =>
+    functionForfactoryS(a1, a2, a3, a4, a5, a6, a7, a8, a9)
+  )
+}
 
 const sequenceTf1 = Apy.sequenceT(E.either)
 

--- a/src/Apply.ts
+++ b/src/Apply.ts
@@ -231,6 +231,49 @@ export function getSemigroup<F, A>(F: Apply<F>, S: Semigroup<A>): () => Semigrou
   })
 }
 
+export interface SequenceT3<F extends URIS3> {
+  <U, L, T extends Array<Type3<F, U, L, any>>>(...t: T & { 0: Type3<F, U, L, any> }): Type3<
+    F,
+    U,
+    L,
+    { [K in keyof T]: [T[K]] extends [Type3<F, U, L, infer A>] ? A : never }
+  >
+}
+export interface SequenceT3C<F extends URIS3, U, L> {
+  <T extends Array<Type3<F, U, L, any>>>(...t: T & { 0: Type3<F, U, L, any> }): Type3<
+    F,
+    U,
+    L,
+    { [K in keyof T]: [T[K]] extends [Type3<F, U, L, infer A>] ? A : never }
+  >
+}
+export interface SequenceT2<F extends URIS2> {
+  <L, T extends Array<Type2<F, L, any>>>(...t: T & { 0: Type2<F, L, any> }): Type2<
+    F,
+    L,
+    { [K in keyof T]: [T[K]] extends [Type2<F, L, infer A>] ? A : never }
+  >
+}
+export interface SequenceT2C<F extends URIS2, L> {
+  <T extends Array<Type2<F, L, any>>>(...t: T & { 0: Type2<F, L, any> }): Type2<
+    F,
+    L,
+    { [K in keyof T]: [T[K]] extends [Type2<F, L, infer A>] ? A : never }
+  >
+}
+export interface SequenceT1<F extends URIS> {
+  <T extends Array<Type<F, any>>>(...t: T & { 0: Type<F, any> }): Type<
+    F,
+    { [K in keyof T]: [T[K]] extends [Type<F, infer A>] ? A : never }
+  >
+}
+export interface SequenceT<F> {
+  <T extends Array<HKT<F, any>>>(...t: T & { 0: HKT<F, any> }): HKT<
+    F,
+    { [K in keyof T]: [T[K]] extends [HKT<F, infer A>] ? A : never }
+  >
+}
+
 const tupleConstructors: { [key: string]: Function1<any, any> } = {}
 
 /**
@@ -247,36 +290,12 @@ const tupleConstructors: { [key: string]: Function1<any, any> } = {}
  *
  * @since 1.5.0
  */
-export function sequenceT<F extends URIS3>(
-  F: Apply3<F>
-): <U, L, T extends Array<Type3<F, U, L, any>>>(
-  ...t: T & { 0: Type3<F, U, L, any> }
-) => Type3<F, U, L, { [K in keyof T]: T[K] extends Type3<F, U, L, infer A> ? A : never }>
-export function sequenceT<F extends URIS3, U, L>(
-  F: Apply3C<F, U, L>
-): <T extends Array<Type3<F, U, L, any>>>(
-  ...t: T & { 0: Type3<F, U, L, any> }
-) => Type3<F, U, L, { [K in keyof T]: T[K] extends Type3<F, U, L, infer A> ? A : never }>
-export function sequenceT<F extends URIS2>(
-  F: Apply2<F>
-): <L, T extends Array<Type2<F, L, any>>>(
-  ...t: T & { 0: Type2<F, L, any> }
-) => Type2<F, L, { [K in keyof T]: T[K] extends Type2<F, L, infer A> ? A : never }>
-export function sequenceT<F extends URIS2, L>(
-  F: Apply2C<F, L>
-): <T extends Array<Type2<F, L, any>>>(
-  ...t: T & { 0: Type2<F, L, any> }
-) => Type2<F, L, { [K in keyof T]: T[K] extends Type2<F, L, infer A> ? A : never }>
-export function sequenceT<F extends URIS>(
-  F: Apply1<F>
-): <T extends Array<Type<F, any>>>(
-  ...t: T & { 0: Type<F, any> }
-) => Type<F, { [K in keyof T]: T[K] extends Type<F, infer A> ? A : never }>
-export function sequenceT<F>(
-  F: Apply<F>
-): <T extends Array<HKT<F, any>>>(
-  ...t: T & { 0: HKT<F, any> }
-) => HKT<F, { [K in keyof T]: T[K] extends HKT<F, infer A> ? A : never }>
+export function sequenceT<F extends URIS3>(F: Apply3<F>): SequenceT3<F>
+export function sequenceT<F extends URIS3, U, L>(F: Apply3C<F, U, L>): SequenceT3C<F, U, L>
+export function sequenceT<F extends URIS2>(F: Apply2<F>): SequenceT2<F>
+export function sequenceT<F extends URIS2, L>(F: Apply2C<F, L>): SequenceT2C<F, L>
+export function sequenceT<F extends URIS>(F: Apply1<F>): SequenceT1<F>
+export function sequenceT<F>(F: Apply<F>): SequenceT<F>
 export function sequenceT<F>(F: Apply<F>): (...args: Array<any>) => HKT<F, any> {
   return (...args: Array<any>) => {
     const len = args.length
@@ -324,28 +343,32 @@ export function sequenceS<F extends URIS3>(
   F: Apply3<F>
 ): <U, L, R extends Record<string, Type3<F, U, L, any>>>(
   r: EnforceNonEmptyRecord<R> & Record<string, Type3<F, U, L, any>>
-) => Type3<F, U, L, { [K in keyof R]: R[K] extends Type3<F, any, any, infer A> ? A : never }>
+) => Type3<F, U, L, { [K in keyof R]: [R[K]] extends [Type3<F, any, any, infer A>] ? A : never }>
 export function sequenceS<F extends URIS3, U, L>(
   F: Apply3C<F, U, L>
 ): <R extends Record<string, Type3<F, U, L, any>>>(
   r: EnforceNonEmptyRecord<R>
-) => Type3<F, U, L, { [K in keyof R]: R[K] extends Type3<F, any, any, infer A> ? A : never }>
+) => Type3<F, U, L, { [K in keyof R]: [R[K]] extends [Type3<F, any, any, infer A>] ? A : never }>
 export function sequenceS<F extends URIS2>(
   F: Apply2<F>
 ): <L, R extends Record<string, Type2<F, L, any>>>(
   r: EnforceNonEmptyRecord<R> & Record<string, Type2<F, L, any>>
-) => Type2<F, L, { [K in keyof R]: R[K] extends Type2<F, any, infer A> ? A : never }>
+) => Type2<F, L, { [K in keyof R]: [R[K]] extends [Type2<F, any, infer A>] ? A : never }>
 export function sequenceS<F extends URIS2, L>(
   F: Apply2C<F, L>
 ): <R extends Record<string, Type2<F, L, any>>>(
   r: EnforceNonEmptyRecord<R>
-) => Type2<F, L, { [K in keyof R]: R[K] extends Type2<F, any, infer A> ? A : never }>
+) => Type2<F, L, { [K in keyof R]: [R[K]] extends [Type2<F, any, infer A>] ? A : never }>
 export function sequenceS<F extends URIS>(
   F: Apply1<F>
 ): <R extends Record<string, Type<F, any>>>(
   r: EnforceNonEmptyRecord<R>
-) => Type<F, { [K in keyof R]: R[K] extends Type<F, infer A> ? A : never }>
-export function sequenceS<F>(F: Apply<F>): (r: Record<string, HKT<F, any>>) => HKT<F, Record<string, any>>
+) => Type<F, { [K in keyof R]: [R[K]] extends [Type<F, infer A>] ? A : never }>
+export function sequenceS<F>(
+  F: Apply<F>
+): <R extends Record<string, HKT<F, any>>>(
+  r: EnforceNonEmptyRecord<R>
+) => HKT<F, { [K in keyof R]: [R[K]] extends [HKT<F, infer A>] ? A : never }>
 export function sequenceS<F>(F: Apply<F>): (r: Record<string, HKT<F, any>>) => HKT<F, Record<string, any>> {
   return r => {
     const keys = Object.keys(r)

